File: app/src/main/java/com/kispoko/tome/util/database/query/UpsertQuery.java

package com.kispoko.tome.util.database.query;


import android.content.ContentValues;
import android.database.sqlite.SQLiteDatabase;

import com.kispoko.tome.Global;

import java.util.UUID;



/**
 * Upsert Query
 */
public class UpsertQuery
{

    // PROPERTIES
    // --------------------------------------------------------------------------------------

    private String        tableName;
    private UUID          rowId;
    private ContentValues row;


    // CONSTRUCTORS
    // --------------------------------------------------------------------------------------

    public UpsertQuery(String tableName, UUID rowId, ContentValues row)
    {
        this.tableName = tableName;
        this.rowId     = rowId;
        this.row       = row;
    }


    // API
    // --------------------------------------------------------------------------------------

    /**
     * Run an upsert query using the Android SQLite library insertWithOnConflict function. If the
     * row doesn't exist, it will be created. We must provide all the values every time though.
     */
    public void run()
    {
        SQLiteDatabase database = Global.getDatabase();

        database.insertWithOnConflict(this.tableName,
                                      null,
                                      this.row,
                                      SQLiteDatabase.CONFLICT_REPLACE);
    }


}

File: app/src/main/java/com/kispoko/tome/lib/orm/sql/query/UpsertQuery.kt

package com.kispoko.tome.lib.orm.sql.query


import android.content.ContentValues
import android.database.sqlite.SQLiteDatabase
import com.kispoko.tome.lib.orm.DatabaseManager
import com.kispoko.tome.lib.orm.ORMLog
import com.kispoko.tome.lib.orm.ORMLogLevel
import com.kispoko.tome.lib.orm.RowInsert
import com.kispoko.tome.lib.orm.sql.SQL
import java.util.*



/**
 * Upsert Query
 */
data class UpsertQuery(val tableName : String, val rowId : UUID, val row : ContentValues)
{

    fun run()
    {
        val database = DatabaseManager.database()

        val validTableName = SQL.validIdentifier(tableName)

        val startTime = System.nanoTime()

        database.insertWithOnConflict(validTableName,
                                      null,
                                      this.row,
                                      SQLiteDatabase.CONFLICT_REPLACE)

        val endTime = System.nanoTime()

        if (ORMLog.logLevel < ORMLogLevel.EVERYTHING)
            ORMLog.event(RowInsert(validTableName, rowId, (endTime - startTime)))
        else
            ORMLog.event(RowInsert(validTableName, rowId, (endTime - startTime), row))
    }


}
